nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java

```diff
@@ -0,0 +1,165 @@
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        service = context.getProperty(ADX_SOURCE_SERVICE).asControllerService(KustoQueryService.class);
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile outgoingFlowFile;
+        String databaseName = context.getProperty(DB_NAME).getValue();
+        String adxQuery;
+        KustoQueryResponse kustoQueryResponse;
+
+        //checks if this processor has any preceding connection, if yes retrieve
+        if (context.hasIncomingConnection()) {
+            FlowFile incomingFlowFile = session.get();
+            //incoming connection exists but the incoming flowfile is null
+            if (incomingFlowFile == null && context.hasNonLoopConnection()) {
+                return;
+            }
+            //incoming connection exists and retrieve adxQuery from context
+            if (incomingFlowFile != null && incomingFlowFile.getSize() == 0) {
+                if (context.getProperty(ADX_QUERY).isSet()) {
+                    adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
+                } else {
+                    String message = "FlowFile query is empty and no scheduled query is set";
+                    getLogger().error(message);
+                    incomingFlowFile = session.putAttribute(incomingFlowFile, ADX_QUERY_ERROR_MESSAGE, message);
+                    session.transfer(incomingFlowFile, FAILED);
+                    return;
+                }
+            } else {
+                try {
+                    adxQuery = getQuery(session, incomingFlowFile);
+                } catch(IOException ioe) {
+                    throw new ProcessException("Failed to read Query from FlowFile",ioe);
+                }
+            }
+            outgoingFlowFile = incomingFlowFile;
+        } else {
+            outgoingFlowFile = session.create();
+            adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(outgoingFlowFile).getValue();
+        }
+
+        //execute Query
+        kustoQueryResponse = executeQuery(databaseName,adxQuery);
+        if(kustoQueryResponse.isError()){
+            //failure scenario
+            getLogger().error("Failed to execute query: {}", kustoQueryResponse.getErrorMessage());
+            outgoingFlowFile = session.putAttribute(outgoingFlowFile, ADX_QUERY_ERROR_MESSAGE, kustoQueryResponse.getErrorMessage());
+            session.transfer(outgoingFlowFile, FAILED);
+        }else {
+            // success scenario
+            session.importFrom(kustoQueryResponse.getAdxQueryResponseStream(), outgoingFlowFile);
```

Comment:
The `ProcessSession.importFrom()` method may not close the provided InputStream, so recommend wrapping this call in try-with-resources:
```suggestion
            try (final InputStream responseStream = kustoQueryResponse.getAdxQueryResponseStream()) {
                session.importFrom(responseStream, outgoingFlowFile);
```